Use conditions for reporting results

krlmlr opened the pull request:
Also: test_that() returns TRUE if successful (no failed expectations, errors, or skips) for programmatic use of testthat.

Bare expectations now simply stop if a test fails. The stop occurs upon the first event, unlike the StopReporter which stops after the first failed test.

krlmlr commented on R/expect-that.r:
```diff
@@ -37,7 +37,18 @@ expect_that <- function(object, condition, info = NULL, label = NULL) {
+                    class = c("test_result", "condition"))
+
+  withRestarts(
+    if (results$passed) signalCondition(cond) else stop(cond),
```
Could also call `message()` instead of `signalCondition()` here.

krlmlr commented:
@hadley @jimhester: PTAL. Part of https://github.com/hadley/testthat/issues/215#issuecomment-75367557.

krlmlr commented on R/test-that.r:
```diff
@@ -73,6 +82,23 @@ test_code <- function(description, code, env) {
+
+  withRestarts(
+    {
+      signalCondition(cond)
```
Call stop() here.

hadley commented on R/test-that.r:
```diff
@@ -73,6 +82,28 @@ test_code <- function(description, code, env) {
   invisible(ok)
 }
 
+report_results <- function(results) {
```
I think this would be better called `report_expectation()`, but I'm not sure `report()` is quite the right verb (but I'm not sure what would be better)

hadley replied:
Or should this all be wrapped up into `expectation()`?

krlmlr replied:
Do you mean that expectation() should create and raise the condition object? This means that we need to be *very* clever in expect_that(): Catch the condition (which is created in the condition(object) call), patch it, and re-raise it. To me, this looks like too much effort for too little gain. But I agree that expectation() should return a proper condition object right away. I'll update the PR soon.

hadley replied:
Maybe we should have `expectation()` (noun) to create the object, and `expect()` (verb) raise the condition?

krlmlr replied:
We already have expect() which accepts a logical and a message; we'll need as.expectation.logical() and as.expectation.expectation() to avoid changing too much unrelated code, but otherwise this is doable. How does that sound?

krlmlr replied:
Then we'll add as.expectation.error(), as.expectation.skip() (as replacement for current expecation_error() and expectation_skip()) and later as.expectation.warning() (#310).

hadley commented on R/test-that.r:
```diff
@@ -73,6 +82,28 @@ test_code <- function(description, code, env) {
+
+raise_condition_from_result <- function(results) {
+  if (results$passed) {
+    cond <- structure(list(results = results),
```
Could we combine this object representation with the expectation, so there's only one type of thing to worry about?

krlmlr commented on R/expectation.r:
```diff
@@ -11,20 +11,73 @@
+as.expectation.error <- function(x, ...) {
+  error <- x$message
+  calls <- x$calls
+  srcref <- x$srcref
```
The srcref handling needs to be improved. This is currently just a placeholder, should be filled later by the calling handler to support file+line reporting.

krlmlr commented:
PTAL: Now expectations are also conditions.

hadley commented on R/expectation.r:
```diff
@@ -11,20 +11,73 @@
+
+new_expectation <- function(failure_msg, success_msg, srcref, ...,
+                            passed = FALSE, error = FALSE, skipped = FALSE) {
+  if (passed) {
```
`passed && !error`?

hadley replied:
Maybe it would be better to have `type = c("error", "failure", "skip", "warning", "message")` etc?

krlmlr replied:
If `error`, then `passed` is `FALSE` by construction (although not enforced). I agree that a type is better, but I'd like to deal with this in a new batch of changes, and also change the reporters -- this PR is already big enough.

hadley replied:
Sure. I think I just introduced some merge conflicts, but if you squash this, I'll fix.

hadley commented on R/test-that.r:
```diff
@@ -51,28 +51,55 @@ test_code <- function(description, code, env) {
   }
+  handle_expectation <- function(exp) {
+    get_reporter()$add_result(exp)
+    if (!isTRUE(exp$passed)) {
```
Maybe pull this out to separate `expectation_ok()` ?

hadley commented:
This is looking really good!

krlmlr commented:
Are you planning further refactorings? Then it would be great to merge this beforehand, otherwise I'll work further on this branch.

hadley commented:
I don't have anything else major planned - I think most of the other issues need this PR completed first.

krlmlr commented:
Extracted expectation_ok(), got rid of success_msg in my code, added two TODO markers.

I think test_code() can be simplified considerably, so that there's basically just one handler function taking care of everything. The reporter should be able to decide if tryCatch() catches the errors or not (#356) :

```
tryCatch(stop(), error = NULL)
tryCatch(stop(), error = function(e) NULL)
```

I'll merge the other PR before adding NEWS to this PR.

hadley commented on R/expectation.r:
```diff
@@ -10,19 +10,76 @@
+  update_expectation(exp, srcref)
+}
+
+update_expectation <- function(exp, srcref, info = NULL, label = NULL) {
```
It feels like we should be able to eventually eliminate this function, wrapping it into the constructor

krlmlr replied:
Difficult without finally deprecating info and label arguments (#218).

hadley replied:
Yeah, lets look again once we've eliminated those args and refactored the expectations

hadley commented on R/test-that.r:
```diff
@@ -52,28 +52,55 @@ test_code <- function(description, code, env) {
   }
+  handle_expectation <- function(exp) {
+    get_reporter()$add_result(exp)
+    if (!expectation_ok(exp)) {
```
`ok <<- expectation_ok(exp)` ?

krlmlr replied:
`ok <<- ok && expectation_ok(exp)` if you prefer.

hadley commented on R/test-that.r:
```diff
@@ -52,28 +52,55 @@ test_code <- function(description, code, env) {
+}
+
+raise_condition <- function(exp) {
+  if (exp$passed) {
```
`expectation_ok(exp)`

hadley commented on tests/testthat/test-return.R:
```diff
@@ -0,0 +1,23 @@
+context("Return values")
```
I think this should probably go in `test_test_that.R`

hadley commented on tests/testthat/test-bare.r:
```diff
@@ -12,3 +12,18 @@ expect_gte(3, 3)
 expect_lt(2, 3)
 expect_lte(2, 2)
+
+expect_error(expect_false(FALSE), NA)
```
I think you accidentally re-added this?

krlmlr replied:
This was intentional, the commented code that was below is now gone.

krlmlr commented:
Done. One more thing: expectation() is exported, you have changed its interface. Perhaps it's safer to keep success_msg (with a warning), or introduce a ... .

hadley commented:
Yeah, I'm not too worried about that. I think the next release will be testthat 1.0.0 so it's ok to make a few breaking changes (and I doubt this one will affect many people)

hadley commented:
And looks good to merge - feel free whenever you think it's ready
